### DISCUSSION TITLE
PHP Unit Tests don't work after updating to Laravel 11

### DISCUSSION THREAD

**Discussion by ErwinLiemburg** [laravel/framework#50797] [Q&A]
### Laravel Version

11

### PHP Version

8.2

### Description

After updating to Laravel 11 I tried to run some PHP Unit tests that I used in the previous Laravel version (10). I use MySql as the main database and for testing I use Sqlite in memory (as set in phpunit.xml). Testing in Laravel 10 worked fine, but running the tests in Laravel 11 showed errors. 

Using MAMP 6.9 for PHP (8.2) and MySql (5.7.39) on Mac M1 with MacOS Sonoma 14.4.1.
Sqlite3 version 3.43.2 installed.

### Steps To Reproduce

I updated a working Laravel 10 project to Laravel 11, following the steps in the Laravel documentation (Prologue > Upgrade guide). Dropped the vendor files, and reinstalled them using composer. At this point the application is still working as supposed to. But running a PHP Unit test

**Testresult:**
`php artisan test --filter AuthorisationTest

   FAIL  Tests\Feature\App\AuthorisationTest
  ⨯ admin can login
  ⨯ teacher can login
  ⨯ student can login
  ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────  
   FAILED  Tests\Feature\App\AuthorisationTest > admin can login                                                                        QueryException   
  SQLSTATE[HY000]: General error: 1 table sqlite_master may not be modified (Connection: sqlite, SQL: delete from sqlite_master where type in ('table', 'index', 'trigger'))

  at vendor/laravel/framework/src/Illuminate/Database/Connection.php:813
    809▕                     $this->getName(), $query, $this->prepareBindings($bindings), $e
    810▕                 );
    811▕             }
    812▕ 
  ➜ 813▕             throw new QueryException(
    814▕                 $this->getName(), $query, $this->prepareBindings($bindings), $e
    815▕             );
    816▕         }
    817▕     }

      +36 vendor frames 
  37  tests/Feature/App/AuthorisationTest.php:19`

Line 19 is the line of parent::setUp();

**AuthorisationTest.php:**
`<?php

namespace Tests\Feature\App;

use App\Enums\Roles;
use App\Models\User;
use App\Providers\RouteServiceProvider;
use Database\Seeders\PermissionsSeeder;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Illuminate\Foundation\Testing\WithFaker;
use Tests\TestCase;

class AuthorisationTest extends TestCase
{
    use RefreshDatabase;

    protected function setUp(): void
    {
        parent::setUp();

        $this->seed(PermissionsSeeder::class);
    }

    public function test_admin_can_login(): void
    {

        $user = User::factory()->create();
        $user->assignRole(Roles::Admin->value);

        $response = $this->post('/login', [
            'email' => $user->email,
            'password' => 'password',
        ]);

        $this->assertAuthenticated();
        $response->assertRedirect(RouteServiceProvider::HOME);
    }

    public function test_teacher_can_login(): void
    {

        $user = User::factory()->create();
        $user->assignRole(Roles::Teacher->value);

        $response = $this->post('/login', [
            'email' => $user->email,
            'password' => 'password',
        ]);

        $this->assertAuthenticated();
        $response->assertRedirect(RouteServiceProvider::HOME);
    }

    public function test_student_can_login(): void
    {

        $user = User::factory()->create();
        $user->assignRole(Roles::Student->value);

        $response = $this->post('/login', [
            'email' => $user->email,
            'password' => 'password',
        ]);

        $this->assertAuthenticated();
        $response->assertRedirect(RouteServiceProvider::HOME);
        $response->assertStatus(302);
    }
}`

**phpunit.xml:**
`<?xml version="1.0" encoding="UTF-8"?>
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/10.4/phpunit.xsd" bootstrap="vendor/autoload.php"
    colors="true">
    <testsuites>
        <testsuite name="Unit">
            <directory suffix="Test.php">./tests/Unit</directory>
        </testsuite>
        <testsuite name="Feature">
            <directory suffix="Test.php">./tests/Feature</directory>
        </testsuite>
    </testsuites>
    <coverage />
    <php>
        <env name="APP_ENV" value="testing" />
        <env name="BCRYPT_ROUNDS" value="4" />
        <env name="CACHE_DRIVER" value="array" />
        <env name="DB_CONNECTION" value="sqlite" />
        <env name="DB_DATABASE" value=":memory:" />
        <env name="MAIL_MAILER" value="array" />
        <env name="QUEUE_CONNECTION" value="sync" />
        <env name="SESSION_DRIVER" value="array" />
        <env name="TELESCOPE_ENABLED" value="false" />
    </php>
    <source>
    <include>
        <directory suffix=".php">./app</directory>
    </include>
    </source>
</phpunit>`

<img width="1087" alt="Screenshot 2024-03-27 at 16 59 36" src="https://github.com/laravel/framework/assets/55760701/762dc871-f9a5-4ef9-aadb-f5c334b0f05a">

**Comment by ZacharyDuBois** [comment]
So, this looks like it is going to be some problem within your application code. Not Laravel. Laravel 11 is a breaking upgrade. Please ensure you follow the upgrade guide [here](https://laravel.com/docs/11.x/upgrade). There were a lot of DB related changes.

**Reply by tutorialsmaterials** [reply]
https://tutorialsmaterial.com/post/new-testing-in-laravel-11-for-beginners
```php

```

**Comment by ErwinLiemburg** [comment]
@ZacharyDuBois Yes, as mentioned I followed the upgrade guide. I will try to do some debugging to find out what is going on to see if it is in my application code.

**Comment by ErwinLiemburg** [comment]
Did some debugging. Looks like the issue occurd when the parent::SetUp() function (in AuthorisationTest.php) is executed. Within that function it tries to refresh the test database. It tries to do a delete `delete from sqlite_master where type in ('table', 'index', 'trigger'))`, but for some reason it crashes. Looks like the sqlite_master table may not be modified.

Look like it's not from the application code, but Laravel code. It could be that I missed something in the upgrade guide, but I'm not sure what I missed.

<img width="672" alt="Screenshot 2024-03-27 at 19 13 46" src="https://github.com/laravel/framework/assets/55760701/1ef539cb-de7e-40ac-927f-bc3631f0c072">

**Comment by crynobone** [comment]
If you're resetting database between tests you should be using `DatabaseMigrations` trait instead of `RefreshDatabase`.

**Reply by FarhanShares** [reply]
Changing to DatabaseMigrations worked for my open-source package [(Laravel Mediaman)](https://github.com/farhanshares/laravel-mediaman). However, it only worked for Laravel 11! Not sure why but you may want to have a look into the [CI Outputs](https://github.com/FarhanShares/laravel-mediaman/actions/runs/8690618069/job/23830969399). Currently I don't need neither of the traits, so the main branch of the repository doesn't have any traits in the tests.

**Comment by ErwinLiemburg** [comment]
@crynobone I replaced RefreshDatabase with DatabaseMigrations. Unfortunately the same result.
Furthermore I'm sorry to hear that this item has been downgraded to a discussion, while in my idea this has to be a bug.

Why can sqlite_master table not be modified? Why does this work in Laravel 10, and not in Laravel 11?

**Reply by crynobone** [reply]
The intended behaviour of `RefreshDatabase` is the only runs the migrations once and reset the database transactions between tests. In-memory database however will "flush" the database between tests by unsetting PDO instance.

This however has been changed (intentionally) for Laravel 11 where PDO instances is being cached between tests.

**Is this a bug?**

No, because the changes was intentionally made and deleting the database between tests is not the intended behaviour of `RefreshDatabase`

**Reply by ErwinLiemburg** [reply]
@crynobone Okay, clear. But why is migrate:fresh crashing while it tries to delete the SQLite_master table before migrating and seeding? This is done in RefreshDatabase as well as DatabaseMigrate.

**Reply by crynobone** [reply]
`migrate:fresh` is only done on the first request and it's not broken even in Laravel Framework own integration tests. Are you sure you are using compatible SQLite versions?

**Reply by ErwinLiemburg** [reply]
I have version `3.43.2 2023-10-10 13:08:14 1b37c146ee9ebb7acd0160c0ab1fd11017a419fa8a3187386ed8cb32b709aapl (64-bit)` installed.

<img width="729" alt="Screenshot 2024-03-28 at 08 25 50" src="https://github.com/laravel/framework/assets/55760701/57f9aec8-70fb-4e24-aba0-f0f96dee668a">

**Reply by crynobone** [reply]
@hafezdivandari What's the support with SQLite and SQLite database with enum column on Laravel 11? https://laracasts.com/discuss/channels/testing/testing-sqlite-vs-mysql This seems to be related to using `enum` column

**Reply by ErwinLiemburg** [reply]
@crynobone To be clear, there is no migrating or seeding going on yet. The only queries, intitiated by db:wipe (FreshCommand.php:46), that are executed are:
`PRAGMA foreign_keys = ON`
`PRAGMA writable_schema = 1`
`delete from sqlite_master where type in ('table', 'index', 'trigger')`

The last query results in: `SQLSTATE[HY000]: General error: 1 table sqlite_master may not be modified`.

According to the sqlite documentation after `PRAGMA writable_schema = 1` you should be able to edit `sqlite_master`.

These actions are not executed in Laravel 10.

**Reply by hafezdivandari** [reply]
@crynobone afaik the support with enum on sqlite has not been changed on Laravel 11, sqlite 3.8.8+ should work when using enum (i.e. check constraint  on varchar).
But this issue may be related to PR #47912?

**Comment by ErwinLiemburg** [comment]
Did some further debugging. Where Laravel 10 had the following code:
`
public function refreshDatabase()
    {
        $this->beforeRefreshingDatabase();

        $this->usingInMemoryDatabase()
                        ? $this->refreshInMemoryDatabase()
                        : $this->refreshTestDatabase();

        $this->afterRefreshingDatabase();
    }

    /**
     * Determine if an in-memory database is being used.
     *
     * @return bool
     */
    protected function usingInMemoryDatabase()
    {
        $default = config('database.default');

        return config("database.connections.$default.database") === ':memory:';
    }

    /**
     * Refresh the in-memory database.
     *
     * @return void
     */
    protected function refreshInMemoryDatabase()
    {
        $this->artisan('migrate', $this->migrateUsing());

        $this->app[Kernel::class]->setArtisan(null);
    }`

Here the function refreshInMemoryDatabase() did a migrate. In Laravel 11 the code changed to the following:
`public function refreshDatabase()
    {
        $this->beforeRefreshingDatabase();

        if ($this->usingInMemoryDatabase()) {
            $this->restoreInMemoryDatabase();
        }

        $this->refreshTestDatabase();

        $this->afterRefreshingDatabase();
    }

    /**
     * Determine if an in-memory database is being used.
     *
     * @return bool
     */
    protected function usingInMemoryDatabase()
    {
        $default = config('database.default');

        return config("database.connections.$default.database") === ':memory:';
    }

    /**
     * Restore the in-memory database between tests.
     *
     * @return void
     */
    protected function restoreInMemoryDatabase()
    {
        $database = $this->app->make('db');

        foreach ($this->connectionsToTransact() as $name) {
            if (isset(RefreshDatabaseState::$inMemoryConnections[$name])) {
                $database->connection($name)->setPdo(RefreshDatabaseState::$inMemoryConnections[$name]);
            }
        }
    }`

For some reason in restoreInMemoryDatabase() there is no migration anymore, variable $name is null, so there is also no $inMemoryConnections available. 

The function refreshTestDatabase() is unchanged.

`protected function refreshTestDatabase()
    {
        if (! RefreshDatabaseState::$migrated) {
            $this->artisan('migrate:fresh', $this->migrateFreshUsing());

            $this->app[Kernel::class]->setArtisan(null);

            RefreshDatabaseState::$migrated = true;
        }

        $this->beginDatabaseTransaction();
    }`

The migrate:fresh command will probably try to delete all tables, but there are no tables due to the fact that no migration has been executed.

Could this be the problem I'm running into? If so, how can it be solved?

**Comment by ErwinLiemburg** [comment]
@crynobone Looks like the problem has been fixed in the latest version I've been updating to (v11.4). I see some issues while testing, but I have to figure out what's creating that (could be an v10 to v11 update issue).
Closing this dicussion.
